### PR TITLE
Dialogue management API

### PR DESCRIPTION
**Pull request by space-pope**
Dialogue management is a subsystem designed to be connected to
NLU results. The dialogue manager processes NLU results and
translates them into system prompts and cues for changes in the
app via the use of a dialogue policy.

Clients may implement their own dialogue policies.

---

This is the first of two PRs, separated so that we can discuss the API itself (which changed a bit during implementation) apart from the initial rule-based policy that makes use of it.

The eventual plan is to wrap all the subsystems in a `Spokestack` class, similar to how the React Native module works, so that configuration is done all at once (maybe through a JSON file or something), and things like `DialogueManager` here are automatically set up to receive NLU events.

The main sticking points for me on this design are the `ConversationData` API and `Prompt`—I'm unsure if they're too idiosyncratic to be effectively extended by end users. `Prompt` in particular seems like it would change based on the dialogue policy in use, but making it into an interface and saving things like reprompts/proposals for the Spokestack policy's implementation had ripple effects through the event API that I didn't like. You end up with weird generics scattered around to avoid typecasting in your dialogue event listener.

Don't be too alarmed by the line count; the boilerplate and javadoc are real.

Also ignore the build failure for now; it's because I intentionally left out `RuleBasedDialoguePolicy` and its friends to make this changeset easier to read, but the dialogue manager sort of relies on it to, um, compile.

### CODE REVIEW

**Review by brentspell — APPROVED** [review]
I don't think I have much feedback here. The code looks cleaner than the elixir version.

File: src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * The current state of the conversation at the time of a dialogue event.
+ */
+public final class ConversationState {
+
+    private final String frameName;
+    private final String nodeName;
+    private final String appAction;
+    private final Prompt systemPrompt;
```
Comment (noelweichbrodt) [inline src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java:15]: 💞 memories 💞

File: src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    private final Map<String, Object> slots;
+
+    /**
+     * Create a new conversation state.
+     *
+     * @param builder  The current conversation state builder.
+     */
+    private ConversationState(Builder builder) {
+        if (builder.conversationNode != null) {
+            String[] parts = builder.conversationNode.split("\\.", 2);
```
Comment (noelweichbrodt) [inline src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java:26]: This is awkward but I understand why it has to be like this

**Review by noelweichbrodt — APPROVED** [review]
Good work here. Much clearer design than previously. I think this is all doable in Swift.

File: src/main/java/io/spokestack/spokestack/dialogue/ConversationData.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+     * Note that since this API is used by Spokestack to store data provided by
+     * the user during voice interactions, {@code key} should not begin with an
+     * underscore ({@code _}) unless intentionally overriding a key set by the
+     * dialogue manager.
+     * </p>
+     *
+     * @param key   The key under which {@code value} will be available.
+     * @param value Data to store for use by the conversation.
+     */
+    void set(String key, Object value);
```
[the remark this answers is not in the document]
Comment (space-pope) [reply src/main/java/io/spokestack/spokestack/dialogue/ConversationData.java:54]: I'm thinking that if clients put custom objects in there, they'll have to track types. Also, as `get` is documented, it's really only for internal use; the main interface that clients need to worry about is `getFormatted`.

I'm not expecting this to be a primary data store for the app, but I did want the dialogue system to be able to store full `Slot` objects in there (though I'll think a bit about whether I truly need that...). I'm open to ideas on how to make this interface more functional/ergonomic. Take a look at how it's used in the policy when I post the follow-up PR and see if anything comes to mind. I think it's a little rough around the edges.

File: src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java (hunk 2 above)
Comment (space-pope) [reply src/main/java/io/spokestack/spokestack/dialogue/ConversationState.java:26]: Yeah, node naming is one of the things I'm on the fence about. It's not super-pretty.

**Comment by space-pope** [comment]
Thanks for the readthroughs and feedback! Followup implementation on the way in a separate PR.

**Pull request merged**